# GWAS for varroa resistance in an heterogeneous honeybee population

<!-- TOC depthFrom:2 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 -->
- [1. Introduction](#1-Introduction)
- [2. Data preparation](#2-Data-preparation)
- [3. Genetic composition estimation and queen genotype reconstruction](#3Genetic-compo-geno)
	- [3.1. ](#31-)
- [4. Phenotypes ](#4-Phenotypes)
- [5. Linkage disequilibrium](#5-LD)
- [6. Simple GWAS](#6-GWAS)
- [7. Meta GWAS](#7-MetaGWAS)
	- [7.1. MANTRA](#71-MANTRA)
	- [7.2. Mash](#72-mash)
- [8. Analysis on genes](#Genes)
<!-- /TOC -->

## 1. Introduction
This repository contains the scripts to perform the analysis described in the paper:
A large genome wide association study on a heterogeneous honeybee, Apis mellifera, population reveals the polygenic genetic architecture for resistance to the major honeybee parasite, Varroa destructor\
Sonia E Eynard, Alain Vignal, Benjamin Basso, Olivier Bouchez, Tabatha Bulach, Yves Le Conte, Benjamin Bainat, Axel Decourtye, Lucie Genestout, Matthieu Guichard, François Guillaume, Emmanuelle Labarthe, Rachid Mahla, Fanny Mondet, Markus, Neudischko, Florence Phocas, Yannick Poquet, Christina Sann, Rémi-Félix Serre, Kamila Tabet, Bertrand Servin \
bioRxiv , doi: \

The genome version used is HAV3.1 (Wallberg et al. 2019, doi: 10.1186/s12864-019-5642-0), the VCF for diversity panel from Wragg et al. 2022 (doi: 10.1111/1755-0998.13665) with 870 and 628 individuals were used. 
All scripts to perform simulations and analysis are available in this GitHub repository, scripts for the statistical models to estimate genetic composition and reconstruct queen genotype are available at the GitHub repository : https://github.com/BertrandServin/beethoven. 
This pipeline is available in the script RUN.sh. 

## 2. Data preparation

## 3. Genetic composition estimation and queen genotype reconstruction (script for the statistical models available at https://github.com/BertrandServin/beethoven)
### 3.1. 

## 4. Phenotypes 


## 5. Linkage disequilibrium

## 6. Simple GWAS

## 7. Meta GWAS
### 7.1. MANTRA
Morris. 2012

### 7.2. Mash
Urbut et al. 2019
 
## 8. Analysis on genes
mBAT-combo (GCTA suite, Li et al. 2022)

